kernel/memory_structs/src/lib.rs

```diff
@@ -531,20 +679,46 @@ macro_rules! implement_page_frame_range {
             
             #[doc = "A `" $TypeName "` that implements `Copy`"]
             #[derive(Clone, Copy)]
-            pub struct [<Copyable $TypeName>] {
-                start: $chunk,
-                end: $chunk,
+            pub struct [<Copyable $TypeName>]<P: PageSize = Page4KiB> {
+                start: $chunk<P>,
+                end: $chunk<P>,
             }
-            impl From<$TypeName> for [<Copyable $TypeName>] {
-                fn from(r: $TypeName) -> Self {
+            impl<P: PageSize + Copy + 'static> From<$TypeName<P>> for [<Copyable $TypeName>]<P> {
+                fn from(r: $TypeName<P>) -> Self {
                     Self { start: *r.start(), end: *r.end() }
                 }
             }
-            impl From<[<Copyable $TypeName>]> for $TypeName {
-                fn from(cr: [<Copyable $TypeName>]) -> Self {
+            impl<P: PageSize + Copy + 'static> From<[<Copyable $TypeName>]<P>> for $TypeName<P> {
+                fn from(cr: [<Copyable $TypeName>]<P>) -> Self {
                     Self::new(cr.start, cr.end)
                 }
             }
+            impl TryFrom<$TypeName> for $TypeName<Page2MiB> {
+                type Error = &'static str;
+                fn try_from(p: $TypeName) -> Result<Self, &'static str> {
+                    if (p.[<size_in_ $chunk:lower s>]() - 1) % 512 == 0 {
```

Comment:
why subtract `1` here? I don't think that's correct.

In fact, I don't think you even need this conditional at all. As long as the `try_from` calls for the `start` and `end` page succeed, that alone is sufficient to ensure the page range can be validly converted into a 2MiB range.